In [1]:
import sys
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
import statsmodels.api as sm

## Thu nhập dữ liệu

In [2]:
vn_house_df = pd.read_csv('VN_housing_dataset.csv')
vn_house_df

Unnamed: 0        Ngày  \
0             0.0  2020-08-05   
1             1.0  2020-08-05   
2             2.0  2020-08-05   
3             3.0  2020-08-05   
4             4.0  2020-08-05   
...           ...         ...   
82492     82492.0  2019-08-07   
82493     82493.0  2019-08-07   
82494     82494.0  2019-08-05   
82495     82495.0  2019-08-05   
82496         NaN         NaN   

                                                 Địa chỉ               Quận  \
0      Đường Hoàng Quốc Việt, Phường Nghĩa Đô, Quận C...      Quận Cầu Giấy   
1      Đường Kim Giang, Phường Kim Giang, Quận Thanh ...    Quận Thanh Xuân   
2      phố minh khai, Phường Minh Khai, Quận Hai Bà T...  Quận Hai Bà Trưng   
3      Đường Võng Thị, Phường Thụy Khuê, Quận Tây Hồ,...        Quận Tây Hồ   
4      Đường Kim Giang, Phường Kim Giang, Quận Thanh ...    Quận Thanh Xuân   
...                                                  ...                ...   
82492  Đường Trần Quốc Hoàn, Phường Quan Hoa, Quận Cầ...      Quận Cầu Giấy   
82493  Đường Nguyễn Khánh Toàn, Phường Quan Hoa, Quận...      Quận Cầu Giấy   
82494  Đường Quan Hoa, Phường Quan Hoa, Quận Cầu Giấy...      Quận Cầu Giấy   
82495  Đường Hồ Tùng Mậu, Phường Mai Dịch, Quận Cầu G...      Quận Cầu Giấy   
82496                                                NaN                NaN   

                  Huyện        Loại hình nhà ở Giấy tờ pháp lý Số tầng  \
0       Phường Nghĩa Đô           Nhà ngõ, hẻm        Đã có sổ       4   
1      Phường Kim Giang  Nhà mặt phố, mặt tiền             NaN     NaN   
2      Phường Minh Khai           Nhà ngõ, hẻm        Đã có sổ       4   
3      Phường Thụy Khuê           Nhà ngõ, hẻm        Đã có sổ     NaN   
4      Phường Kim Giang           Nhà ngõ, hẻm             NaN     NaN   
...                 ...                    ...             ...     ...   
82492   Phường Quan Hoa  Nhà mặt phố, mặt tiền             NaN     NaN   
82493   Phường Quan Hoa  Nhà mặt phố, mặt tiền        Đã có sổ     NaN   
82494   Phường Quan Hoa           Nhà ngõ, hẻm        Đã có sổ     NaN   
82495   Phường Mai Dịch        Nhà phố liền kề             NaN     NaN   
82496               NaN                    NaN             NaN     NaN   

      Số phòng ngủ Diện tích      Dài Rộng           Giá/m2  
0          5 phòng     46 m²      NaN  NaN   86,96 triệu/m²  
1          3 phòng     37 m²      NaN  NaN  116,22 triệu/m²  
2          4 phòng     40 m²     10 m  4 m      65 triệu/m²  
3          6 phòng     51 m²  12.75 m  4 m     100 triệu/m²  
4          4 phòng     36 m²      9 m  4 m   86,11 triệu/m²  
...            ...       ...      ...  ...              ...  
82492      3 phòng     50 m²      NaN  NaN     292 triệu/m²  
82493      4 phòng     41 m²      NaN  NaN  341,46 triệu/m²  
82494      4 phòng     60 m²      NaN  NaN  101,67 triệu/m²  
82495      4 phòng     45 m²      NaN  NaN  102,22 triệu/m²  
82496          NaN       NaN      NaN  NaN              NaN  

[82497 rows x 13 columns]

### Có bao nhiêu dòng cột? Có dòng nào bị lặp không ?

In [3]:
vn_house_df.shape

(82497, 13)

In [4]:
have_duplicated_rows = vn_house_df.duplicated().any()
have_duplicated_rows

False

## Khám phá dữ liệu

In [5]:
vn_house_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 82497 entries, 0 to 82496
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Unnamed: 0       82496 non-null  float64
 1   Ngày             82496 non-null  object 
 2   Địa chỉ          82449 non-null  object 
 3   Quận             82495 non-null  object 
 4   Huyện            82449 non-null  object 
 5   Loại hình nhà ở  82465 non-null  object 
 6   Giấy tờ pháp lý  53610 non-null  object 
 7   Số tầng          36399 non-null  object 
 8   Số phòng ngủ     82458 non-null  object 
 9   Diện tích        82495 non-null  object 
 10  Dài              19827 non-null  object 
 11  Rộng             35445 non-null  object 
 12  Giá/m2           82484 non-null  object 
dtypes: float64(1), object(12)
memory usage: 8.2+ MB


In [6]:
vn_house_df.nunique()

Unnamed: 0         82496
Ngày                 174
Địa chỉ            13957
Quận                  29
Huyện                309
Loại hình nhà ở        4
Giấy tờ pháp lý        3
Số tầng               29
Số phòng ngủ          11
Diện tích            504
Dài                  487
Rộng                 410
Giá/m2              8345
dtype: int64

Như vậy, ta có thể nhận định được các cột categorical và numerical một cách dễ dàng.

In [7]:
categorical_cols = ['Loại hình nhà ở','Giấy tờ pháp lý']
numerical_cols = ['Số tầng','Số phòng ngủ','Diện tích','Dài','Rộng','Giá/m2']

Lưu ý:
- Cột Ngày ta sẽ không xét bởi vì không có giá trị thực tiễn.
- Cột Quận, Huyện cũng không xét bởi vì thực chất Quận với Huyện đã nằm ở trong cột Địa Chỉ.
- Cột Dài, Rộng không xét bởi vì cột Diện tích = Dài x Rộng.

## Tiền xử lý dữ liệu

Nhìn vào dữ liệu, ta cần lưu ý vài thứ như sau:
- Ta đang xét tập dữ liệu về nhà ở có các biến về thông tin ngôi nhà mà những nhà này không được gắn id hoặc bất cứ biến nào phù hợp để phân biệt ngoại trừ địa chỉ. Vì thế, cột Ngày không có giá trị trực quan cũng như phân tích. Ta sẽ bỏ cột này.
- Nếu nhìn sâu một chút, cột Quận Huyện đã nằm bên trong cột địa chỉ bởi vì địa chỉ được ghi một cách đầy đủ theo quy định ở nước ta nên ta sẽ bỏ luôn hai cột Quận, Huyện.
- Cột Diện tích thực chất là tích của hai cột Dài và Rộng. Như vậy, ta cân nhắc bỏ hai cột này.

In [8]:
removed_cols = ['Unnamed: 0','Ngày','Quận','Huyện','Dài','Rộng']
vn_house_df.drop(columns=removed_cols,inplace=True)
vn_house_df

Địa chỉ  \
0      Đường Hoàng Quốc Việt, Phường Nghĩa Đô, Quận C...   
1      Đường Kim Giang, Phường Kim Giang, Quận Thanh ...   
2      phố minh khai, Phường Minh Khai, Quận Hai Bà T...   
3      Đường Võng Thị, Phường Thụy Khuê, Quận Tây Hồ,...   
4      Đường Kim Giang, Phường Kim Giang, Quận Thanh ...   
...                                                  ...   
82492  Đường Trần Quốc Hoàn, Phường Quan Hoa, Quận Cầ...   
82493  Đường Nguyễn Khánh Toàn, Phường Quan Hoa, Quận...   
82494  Đường Quan Hoa, Phường Quan Hoa, Quận Cầu Giấy...   
82495  Đường Hồ Tùng Mậu, Phường Mai Dịch, Quận Cầu G...   
82496                                                NaN   

             Loại hình nhà ở Giấy tờ pháp lý Số tầng Số phòng ngủ Diện tích  \
0               Nhà ngõ, hẻm        Đã có sổ       4      5 phòng     46 m²   
1      Nhà mặt phố, mặt tiền             NaN     NaN      3 phòng     37 m²   
2               Nhà ngõ, hẻm        Đã có sổ       4      4 phòng     40 m²   
3               Nhà ngõ, hẻm        Đã có sổ     NaN      6 phòng     51 m²   
4               Nhà ngõ, hẻm             NaN     NaN      4 phòng     36 m²   
...                      ...             ...     ...          ...       ...   
82492  Nhà mặt phố, mặt tiền             NaN     NaN      3 phòng     50 m²   
82493  Nhà mặt phố, mặt tiền        Đã có sổ     NaN      4 phòng     41 m²   
82494           Nhà ngõ, hẻm        Đã có sổ     NaN      4 phòng     60 m²   
82495        Nhà phố liền kề             NaN     NaN      4 phòng     45 m²   
82496                    NaN             NaN     NaN          NaN       NaN   

                Giá/m2  
0       86,96 triệu/m²  
1      116,22 triệu/m²  
2          65 triệu/m²  
3         100 triệu/m²  
4       86,11 triệu/m²  
...                ...  
82492     292 triệu/m²  
82493  341,46 triệu/m²  
82494  101,67 triệu/m²  
82495  102,22 triệu/m²  
82496              NaN  

[82497 rows x 7 columns]

Ta sẽ cập nhật lại numerical_cols

In [9]:
numerical_cols = ['Số tầng','Số phòng ngủ','Diện tích','Giá/m2']

Tiếp theo, ta cần đổi các dạng dữ liệu ở các cột numerical sang số thay vì object. Ví dụ, cột "Số phòng ngủ" dư chữ phòng, ta sẽ muốn giữ lại số thôi. Tương tự với "Diện tích" và "Giá/m2" loại bỏ các đơn vị đo không cần thiết.

Tuy nhiên, ta có các giá trị như đ/m2, tỷ/m2, ta sẽ bỏ qua các giá trị này bởi vì phần lớn dữ liệu tập trung ở triệu/m2.

Ta vẫn có thể chuyển đổi giá trị tuy nhiên có thể khiến dữ liệu xuất hiện các outlier bởi vì tỷ/m2 quá lớn trong khi đ/m2 quá nhỏ.

In [10]:
vn_house_df = vn_house_df[vn_house_df['Giá/m2'].str.contains('triệu/m²') == True]
vn_house_df

Địa chỉ  \
0      Đường Hoàng Quốc Việt, Phường Nghĩa Đô, Quận C...   
1      Đường Kim Giang, Phường Kim Giang, Quận Thanh ...   
2      phố minh khai, Phường Minh Khai, Quận Hai Bà T...   
3      Đường Võng Thị, Phường Thụy Khuê, Quận Tây Hồ,...   
4      Đường Kim Giang, Phường Kim Giang, Quận Thanh ...   
...                                                  ...   
82491  Đường Hồ Tùng Mậu, Phường Phúc Diễn, Quận Bắc ...   
82492  Đường Trần Quốc Hoàn, Phường Quan Hoa, Quận Cầ...   
82493  Đường Nguyễn Khánh Toàn, Phường Quan Hoa, Quận...   
82494  Đường Quan Hoa, Phường Quan Hoa, Quận Cầu Giấy...   
82495  Đường Hồ Tùng Mậu, Phường Mai Dịch, Quận Cầu G...   

             Loại hình nhà ở Giấy tờ pháp lý Số tầng Số phòng ngủ Diện tích  \
0               Nhà ngõ, hẻm        Đã có sổ       4      5 phòng     46 m²   
1      Nhà mặt phố, mặt tiền             NaN     NaN      3 phòng     37 m²   
2               Nhà ngõ, hẻm        Đã có sổ       4      4 phòng     40 m²   
3               Nhà ngõ, hẻm        Đã có sổ     NaN      6 phòng     51 m²   
4               Nhà ngõ, hẻm             NaN     NaN      4 phòng     36 m²   
...                      ...             ...     ...          ...       ...   
82491        Nhà phố liền kề             NaN     NaN      3 phòng     38 m²   
82492  Nhà mặt phố, mặt tiền             NaN     NaN      3 phòng     50 m²   
82493  Nhà mặt phố, mặt tiền        Đã có sổ     NaN      4 phòng     41 m²   
82494           Nhà ngõ, hẻm        Đã có sổ     NaN      4 phòng     60 m²   
82495        Nhà phố liền kề             NaN     NaN      4 phòng     45 m²   

                Giá/m2  
0       86,96 triệu/m²  
1      116,22 triệu/m²  
2          65 triệu/m²  
3         100 triệu/m²  
4       86,11 triệu/m²  
...                ...  
82491   81,58 triệu/m²  
82492     292 triệu/m²  
82493  341,46 triệu/m²  
82494  101,67 triệu/m²  
82495  102,22 triệu/m²  

[81626 rows x 7 columns]

In [11]:
def remove(x):
    floor = "Nhiều hơn 10"
    room = "phòng"
    other_room = "nhiều hơn 10 phòng"
    meter = "m²"
    milion_meter = " triệu/m²"
    if floor in x:
        x = int(x.replace(floor,'11'))
    elif other_room in x:
        x = 11
    elif room in x:
        x = int(x.replace(room,""))
    elif milion_meter in x:
        x = x.replace(",","")
        x = float((x.replace(milion_meter,"")))
    elif meter in x:
        x = float(x.replace(meter,""))
    return x

In [12]:
modified_cols = ['Số tầng','Số phòng ngủ','Diện tích','Giá/m2']
for col in modified_cols:
    data = vn_house_df[col].dropna()
    vn_house_df[col] = data.apply(remove)
vn_house_df['Số tầng'] = pd.to_numeric(vn_house_df['Số tầng'])

C:\Users\TCTOAN\AppData\Local\Temp/ipykernel_10704/3871185674.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  vn_house_df[col] = data.apply(remove)
C:\Users\TCTOAN\AppData\Local\Temp/ipykernel_10704/3871185674.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  vn_house_df['Số tầng'] = pd.to_numeric(vn_house_df['Số tầng'])


In [13]:
vn_house_df

Địa chỉ  \
0      Đường Hoàng Quốc Việt, Phường Nghĩa Đô, Quận C...   
1      Đường Kim Giang, Phường Kim Giang, Quận Thanh ...   
2      phố minh khai, Phường Minh Khai, Quận Hai Bà T...   
3      Đường Võng Thị, Phường Thụy Khuê, Quận Tây Hồ,...   
4      Đường Kim Giang, Phường Kim Giang, Quận Thanh ...   
...                                                  ...   
82491  Đường Hồ Tùng Mậu, Phường Phúc Diễn, Quận Bắc ...   
82492  Đường Trần Quốc Hoàn, Phường Quan Hoa, Quận Cầ...   
82493  Đường Nguyễn Khánh Toàn, Phường Quan Hoa, Quận...   
82494  Đường Quan Hoa, Phường Quan Hoa, Quận Cầu Giấy...   
82495  Đường Hồ Tùng Mậu, Phường Mai Dịch, Quận Cầu G...   

             Loại hình nhà ở Giấy tờ pháp lý  Số tầng  Số phòng ngủ  \
0               Nhà ngõ, hẻm        Đã có sổ      4.0           5.0   
1      Nhà mặt phố, mặt tiền             NaN      NaN           3.0   
2               Nhà ngõ, hẻm        Đã có sổ      4.0           4.0   
3               Nhà ngõ, hẻm        Đã có sổ      NaN           6.0   
4               Nhà ngõ, hẻm             NaN      NaN           4.0   
...                      ...             ...      ...           ...   
82491        Nhà phố liền kề             NaN      NaN           3.0   
82492  Nhà mặt phố, mặt tiền             NaN      NaN           3.0   
82493  Nhà mặt phố, mặt tiền        Đã có sổ      NaN           4.0   
82494           Nhà ngõ, hẻm        Đã có sổ      NaN           4.0   
82495        Nhà phố liền kề             NaN      NaN           4.0   

       Diện tích   Giá/m2  
0           46.0   8696.0  
1           37.0  11622.0  
2           40.0     65.0  
3           51.0    100.0  
4           36.0   8611.0  
...          ...      ...  
82491       38.0   8158.0  
82492       50.0    292.0  
82493       41.0  34146.0  
82494       60.0  10167.0  
82495       45.0  10222.0  

[81626 rows x 7 columns]

Tiền xử lý hoàn thành

## Thống kê mô tả

### Võ Quang Tuấn

## Phân tích dữ liệu

### Trương Chí Toàn

### Với mỗi cột numerical, dữ liệu được phân bố như thế nào?

In [21]:
def missing_ratio(col):
    return col.isna().mean() * 100
nume_col_profiles_df = vn_house_df[numerical_cols].agg([missing_ratio,pd.Series.min,pd.Series.max])
nume_col_profiles_df

Số tầng  Số phòng ngủ  Diện tích   Giá/m2
missing_ratio  55.833925      0.046554        0.0      0.0
min             1.000000      1.000000        2.0      1.0
max            73.000000     11.000000    10360.0  98361.0

Như ta thấy, số lượng min max của mỗi cột đều khá hợp lý, nên ta sẽ xét lượng missing_ratio

Cột diện tích và Giá/m2 không bị thiếu dữ liệu. Số phòng ngủ thì chỉ thiếu khoảng 0.05%, không đáng kể. Điều bất thường ở đây là số tầng thì lại thiếu tới 55,8% tới hơn 1 nửa tổng dữ liệu. Khá là kỳ lạ

### Với mỗi cột categorical, dữ liệu được phân bố như thế nào?

In [19]:
def num_diff_vals(col):
    return len(col.value_counts())
def diff_vals(col):
    return col.value_counts().index.unique().to_numpy()

cate_col_profiles_df = vn_house_df[categorical_cols].agg([missing_ratio,num_diff_vals,diff_vals])
cate_col_profiles_df

Loại hình nhà ở  \
missing_ratio                                           0.035528   
num_diff_vals                                                  4   
diff_vals      [Nhà ngõ, hẻm, Nhà mặt phố, mặt tiền, Nhà phố ...   

                                     Giấy tờ pháp lý  
missing_ratio                              35.070933  
num_diff_vals                                      3  
diff_vals      [Đã có sổ, Đang chờ sổ, Giấy tờ khác]

Loại hình nhà ở có 4 giá trị khác nhau là Nhà ngõ, hẻm, Nhà mặt phố, mặt tiền, Nhà phố liền kề.

Giấy tờ pháp lý có 3 giá trị khác nhau là Đã có sổ, Đang chờ sổ, Giấy tờ khác

Xét về missing ratio thì loại hình nhà ở chỉ thiếu khoảng 0.03%, không đáng kể, trong khi giấy tờ pháp lý thì lại thiếu tới 35%. Chứng tỏ khoảng 1/3 hộ gia đình thiếu hoặc không khai báo giấy tờ nhà của hộ.

## Phân tích hồi quy, dự báo

### Trần Quỳnh Như